src/axom/primal/geometry/NURBSPatch.hpp

```diff
@@ -0,0 +1,2745 @@
+    evaluateDerivatives(u, v, 1, ders);
+
+    return ders[1][0];
+  }
+
+  /*!
+   * \brief Computes a tangent in v of the NURBS patch at (\a u, \a v)
+   *
+   * \param [in] u Parameter value at which to evaluate on the first axis
+   * \param [in] v Parameter value at which to evaluate on the second axis
+   *
+   * \pre We require evaluation of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dv(T u, T v) const
+  {
+    axom::Array<VectorType, 2> ders;
+    evaluateDerivatives(u, v, 1, ders);
+
+    return ders[0][1];
+  }
+
+  /*!
+   * \brief Computes the second derivative in u of a NURBS patch at (\a u, \a v)
+   * 
+   * \param [in] u Parameter value at which to evaluate on the first axis
+   * \param [in] v Parameter value at which to evaluate on the second axis
+   * 
+   * \pre We require evaluation of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dudu(T u, T v) const
+  {
+    axom::Array<VectorType, 2> ders;
+    evaluateDerivatives(u, v, 2, ders);
+
+    return ders[2][0];
+  }
+
+  /*!
+   * \brief Computes the second derivative in v of a NURBS patch at (\a u, \a v)
+   * 
+   * \param [in] u Parameter value at which to evaluate on the first axis
+   * \param [in] v Parameter value at which to evaluate on the second axis
+   * 
+   * \pre We require evaluation of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dvdv(T u, T v) const
+  {
+    axom::Array<VectorType, 2> ders;
+    evaluateDerivatives(u, v, 2, ders);
+
+    return ders[0][2];
+  }
+
+  /*!
+   * \brief Computes the mixed second derivative in u and v of a NURBS patch at (\a u, \a v)
+   * 
+   * \param [in] u Parameter value at which to evaluate on the first axis
+   * \param [in] v Parameter value at which to evaluate on the second axis
+   * 
+   * \pre We require evaluation of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dudv(T u, T v) const
+  {
+    axom::Array<VectorType, 2> ders;
+    evaluateDerivatives(u, v, 2, ders);
+
+    return ders[1][1];
+  }
+
+  /*!
+   * \brief Computes the mixed second derivative in u and v of a NURBS patch at (\a u, \a v)
+   * 
+   * \param [in] u Parameter value at which to evaluate on the first axis
+   * \param [in] v Parameter value at which to evaluate on the second axis
+   * 
+   * \pre We require evaluation of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dvdu(T u, T v) const { return dudv(u, v); }
+
+  /*!
+   * \brief Computes the normal vector to the NURBS patch at (\a u, \a v)
+   * 
+   * \param [in] u Parameter value at which to evaluate on the first axis
+   * \param [in] v Parameter value at which to evaluate on the second axis
+   * 
+   * \pre We require evaluation of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType normal(T u, T v) const
+  {
+    PointType eval;
+    VectorType Du, Dv;
+    evaluateFirstDerivatives(u, v, eval, Du, Dv);
+
+    return VectorType::cross_product(Du, Dv);
```

Comment:
Are there cases where the derivative is not defined, e.g. near cusps? If so, what will the normal be?

Reply:
It's possible that one or both derivatives are zero, in which case the normal would be zero too. That doesn't mean that the surface itself doesn't have a normal there (like for a corner of a degenerate triangular patch), but the normal with respect to this *parameterization* is zero. Accounting for all the relevant edge cases is tricky, and I haven't worked out a consistent system for that yet.

Reply:
Would it make sense to add a note? Or perhaps later when the edge cases are more worked out?

Reply:
Couldn't hurt to add a comment here, although it isn't like returning a zero vector is itself incorrect. For a simple cusp on a 2D curve, the tangent vector at the point of the cusp really is just continuously zero, and although I have less intuition for the 3D case, I'm fairly sure that's the case there too. Even though the direction of the tangent/normal vector changes depending on the direction of approach, the magnitude goes to zero regardless. In my experience, the more pressing issue is when you need to take a *unit* normal vector at a point like that, since that genuinely is always going to be discontinuous at the cusp. This is a bit troublesome with the current `primal::Vector` implementation, which returns something like `(1,0,0)` by default if the norm is zero.

As an example of how I've dealt with these geometric/unit vectors before, the 2D winding number classes need to know tangent vectors at endpoints, and in those cases I manually iterated over control points until I found a pair that correctly defined a geometric tangent vector. That doesn't strictly help with the cusp case, but in 2D you could imagine splitting the curve at the point of discontinuity and handling each half that way.